### PR TITLE
HTTPCORE-589 Fix LaxConnPool to ensure pending requests assigned connections

### PR DESCRIPTION
**Pull request by coxlinton**
from available pool

### CODE REVIEW

File: httpcore5/src/main/java/org/apache/hc/core5/pool/LaxConnPool.java [hunk 1]
```diff
@@ -392,17 +395,17 @@ public void shutdown(final CloseMode closeMode) {
...
             do {
                 prev = allocated.get();
                 next = (prev<poolmax)? prev+1 : prev;
             } while (!allocated.compareAndSet(prev, next));
-            return prev < next;
+            return (prev < next)? new PoolEntry<T,C>(route, timeToLive) : null;
         }
 
-        private void deallocatePoolEntry() {
+        private void disposePoolEntry(final PoolEntry<T, C> entry) {
```
Comment (ok2c) [inline httpcore5/src/main/java/org/apache/hc/core5/pool/LaxConnPool.java]: @coxlinton The `entry` parameter does not appear to be used and `disposePoolEntry` does not seem right given that no disposal actually takes place in this method. Can you think of a better name for this method?

File: httpcore5/src/main/java/org/apache/hc/core5/pool/LaxConnPool.java [hunk 2]
```diff
@@ -481,12 +485,18 @@ public void release(final PoolEntry<T, C> releasedEntry, final boolean reusable)
...
+            releaseSeqNum.incrementAndGet();
             servicePendingRequest();
         }
 
+
         private void servicePendingRequest() {
+            servicePendingRequests(false);
+        }
+
+        private void servicePendingRequests(final boolean wholeQueue) {
```
Comment (ok2c) [inline httpcore5/src/main/java/org/apache/hc/core5/pool/LaxConnPool.java]: @coxlinton Let's use an enum here instead of a boolean.

Comment (garydgregory) [reply httpcore5/src/main/java/org/apache/hc/core5/pool/LaxConnPool.java]: - enum  QueueProcessingPolicy? QueueProcessingStrategy?
- PARTIAL vs. COMPLETE? 
- COMPLETE vs. INCOMPLETE?
- WHOLE vs. PARTIAL?

File: httpcore5/src/main/java/org/apache/hc/core5/pool/LaxConnPool.java (hunk 1 above)
Comment (coxlinton) [reply httpcore5/src/main/java/org/apache/hc/core5/pool/LaxConnPool.java]: I intentionally created this method to be the matched pair to the concept of createPoolEntry: one method creates a new pool entry for the pool, the other is used for "the reverse of that"; the right description of that it the tricky thing :) Although the entry param currently isn't used, imagine a modification of the pool structure where we maintain a master list of all pool entries: then createPoolEntry would add the new entry to the master list, and disposePoolEntry would remove the entry from the master list. Would discardPoolEntry be any better?
But perhaps that's all unnecessary and I could just leave it as the previously declared deallocatePoolEntry - it's not really quite the exact analogue of createPoolEntry, but does accurately describe the current function of the method.

Comment (ok2c) [reply httpcore5/src/main/java/org/apache/hc/core5/pool/LaxConnPool.java]: @coxlinton I half suspected that as being the reason. I fully understand your desire to have a pair of symmetric allocation / deallocation methods. However, at present it just makes things more confusing. Those methods are not API visible and can be changed at point it becomes warranted. For now I would even go as far as removing `disposePoolEntry` and calling `allocated.decrementAndGet()` directly.

Comment (coxlinton) [reply httpcore5/src/main/java/org/apache/hc/core5/pool/LaxConnPool.java]: I've returned the method to it's former definition, and squashed the commits, so hopefully this is ready to merge now.

**Pull request merged**